### ISSUE TITLE
Create standardised Adapter structure and API

### ISSUE THREAD

**Issue by Mearman** [breadboard-ai/breadboard#905]

**Comment by dglazkov** [comment]
A naive sketch: an adapter is a small npm package that one can use to quickly bootstrap running a board in a given environment.

- Slack: create a slack app that loads and runs a board
- Discord: create a Discord app that loads and runs a board
- React: a ready-made component that loads and runs a board

WDYT?

**Comment by Mearman** [comment]
Thinking about implementation it feels like most of the hurdles when implementing each would be asynchronicity of inputs. I came up with a somewhat similar solution when we were working on the DevPulse sample. 

[ControllableAsyncGeneratorRunner.ts](https://github.com/ExaDev-io/breadboard-samples/blob/develop/samples/DevPulse%20V4/src/lib/classes/ControllableAsyncGeneratorRunner.ts)

What if we were to split this into an interface and/or abstract class?

**Comment by dglazkov** [comment]
Aha! That's the kind of stuff we want to look at. Why did you end up building it? Let's see if we can suss out the API from this.

**Comment by dglazkov** [comment]
> asynchronicity of inputs

probably there somewhere, but I'd love for you to expand.

**Comment by dglazkov** [comment]
What would be the ideal API interaction model here? It sounds like maybe:

- give me the inputs and outputs before running
- when I give you inputs, start running
- give me outputs.

**Comment by Mearman** [comment]
@dglazkov 
> What would be the ideal API interaction model here? It sounds like maybe:
> 
> * give me the inputs and outputs before running

Do you mean on a board, node or per-iteration level? Tell me if I'm wrong, but the execution of each node will influence its outputs and, as a result, the inputs that are fulfilled on the following nodes. So, ahead of time, we can say with certainty which nodes will always need an input, but not which will not, or will only need it some of the time.

> * when I give you inputs, start running

^ do you mean unpause/continue the execution? 

> * give me outputs.
> 
> There's something about handling cycles though.

I think my point above alludes to this. As in, if the board has multiple iterations, the inputs and outputs may vary.
 
> Maybe it's more like an even listener API?
> 
> * subscribe to input events and output events by providing a handler for each
> * run board

^ This is sort of where I'm thinking it's going, too. And the [ControllableAsyncGeneratorRunner.ts](https://github.com/ExaDev-io/breadboard-samples/blob/develop/samples/DevPulse%20V4/src/lib/classes/ControllableAsyncGeneratorRunner.ts) somewhat goes towards solving it. There is a similar usage in Breadboard Web that @TinaNikou is digging out.

**Comment by Mearman** [comment]
@dglazkov what were your thoughts on this?
> Do you mean on a board, node or per-iteration level? Tell me if I'm wrong, but the execution of each node will influence its outputs and, as a result, the inputs that are fulfilled on the following nodes. So, ahead of time, we can say with certainty which nodes will always need an input, but not which will not, or will only need it some of the time.

**Comment by dglazkov** [comment]
> @dglazkov what were your thoughts on this?
> 
> > Do you mean on a board, node or per-iteration level? Tell me if I'm wrong, but the execution of each node will influence its outputs and, as a result, the inputs that are fulfilled on the following nodes. So, ahead of time, we can say with certainty which nodes will always need an input, but not which will not, or will only need it some of the time.

Yep agreeing on that -- and leaning toward events-based API. Naively there are going to be two use cases:
1. I have a board that I want to run as part of an app. Here, I will likely know all inputs/outputs pretty well, even if the board has cycles.
2.  I want to build an app that runs an arbitrary board. This one is going to be more tricky.

I am thinking maybe let's optimize for the first one?

**Comment by Mearman** [comment]
I think for web-based things, this may require [`breadboard-wed` to come of age](https://github.com/breadboard-ai/breadboard/issues/206). Though my investigations from #958 make me think it may be required for non-web-based compatibility too.

@alexandraM98 what do you think regarding @dglazkov use-cases listed above?

**Comment by jamesjacko** [comment]
Our current POA on this involves 3 initial packages each dependent on the previous package:

1. Core adapter package - provides an abstracted pattern for interacting and using breadboard in a JS application. This exposes an API for interacting with the generators of the breadboard.
2. React adapter package - provide a react context, provider and hooks to interact with a board or boards via the core package. The board shares the same lifetime of the component that the provider is wrapped around. In most cases, this will be the main <App /> and providing routing happens within the component, the components will always interact with the same instance of the board(s) unless explicitly told otherwise.
3. React component library - library of wrapped lit components which can be used in a react application, this library will also offer the core and react adapter packages via dependencies (may end up as peer dependencies).

**Comment by dglazkov** [comment]
sounds great! Please share early and let's iterate.

**Comment by dglazkov** [comment]
This might be relevant: https://breadboard-ai.github.io/breadboard/docs/inspector/run/
